stripe/src/main/java/com/stripe/android/AbstractEphemeralKey.java

```diff
@@ -81,6 +85,66 @@ private EphemeralKey(
         mType = type;
     }
 
+    protected AbstractEphemeralKey(
+            @Nullable JSONObject jsonObject
+    ) throws JSONException {
+        mCreated = jsonObject.getLong(FIELD_CREATED);
+        mExpires = jsonObject.getLong(FIELD_EXPIRES);
+        mId = jsonObject.getString(FIELD_ID);
+        mLiveMode = jsonObject.getBoolean(FIELD_LIVEMODE);
+        mObject = jsonObject.getString(FIELD_OBJECT);
+        mSecret = jsonObject.getString(FIELD_SECRET);
+
+        // Get the values from the associated objects array first element
+        JSONArray associatedObjectArray = jsonObject.getJSONArray(FIELD_ASSOCIATED_OBJECTS);
+        JSONObject typeObject = associatedObjectArray.getJSONObject(0);
+        mType = typeObject.getString(FIELD_TYPE);
+        mObjectId = typeObject.getString(FIELD_ID);
+    }
+
+    @Nullable
+    protected static <TEphemeralKey extends AbstractEphemeralKey> TEphemeralKey
+    fromString(@Nullable String rawJson, Class ephemeralKeyClass) {
+        if (rawJson == null) {
+            return null;
+        }
+
+        try {
+            JSONObject object = new JSONObject(rawJson);
+            return fromJson(object, ephemeralKeyClass);
+        } catch (JSONException ignored) {
+            return null;
+        }
+    }
+
+    @Nullable
+    @SuppressWarnings({"checkstyle:LineLength", "checkstyle:IllegalCatch"})
+    protected static <TEphemeralKey extends AbstractEphemeralKey> TEphemeralKey
+    fromJson(@Nullable JSONObject jsonObject, Class ephemeralKeyClass) {
+        if (jsonObject == null) {
+            return null;
+        }
+
+        try {
+            return (TEphemeralKey)
+                    ephemeralKeyClass.getConstructor(JSONObject.class).newInstance(jsonObject);
```

Comment:
Also it avoids the weird code with the if statements that was there before, with a tad bit of reflection